### PR TITLE
Dotnet tool restore in parallel

### PR DESCRIPTION
**Pull request by wli3**
Need https://github.com/dotnet/cli/pull/10246 get in first due to merge conflict

Make dotnet tool restore run in parallel so I will be faster when there are multiple tools

### CODE REVIEW

File: src/dotnet/commands/dotnet-tool/restore/ToolRestoreCommand.cs [hunk 1]
```diff
@@ -96,81 +98,110 @@ public override int Execute()
                 return 0;
             }
 
-            Dictionary<RestoredCommandIdentifier, RestoredCommand> dictionary =
```
Comment (wli3) [inline src/dotnet/commands/dotnet-tool/restore/ToolRestoreCommand.cs]: Please review with split diff. The github is confused so bad. I made all "output" collection to be concurrent version. And run an extract method to make it obvious that all the input and output are either readonly or concurrent collection

**Comment by wli3** [comment]
@dotnet-bot test RHEL6 x64 Debug Build

**Comment by wli3** [comment]
@dotnet-bot test Ubuntu x64 Release Build

**Comment by wli3** [comment]
@dotnet/dotnet-cli ping review

**Comment by wli3** [comment]
@dotnet-bot test .NET Core CLI CI

File: src/dotnet/commands/dotnet-tool/restore/ToolRestoreCommand.cs [hunk 2]
```diff
@@ -96,117 +97,111 @@ public override int Execute()
...
-                new Dictionary<RestoredCommandIdentifier, RestoredCommand>();
+            ToolRestoreResult[] toolRestoreResults =
+                packagesFromManifest
+                    .Select(package => InstallPackages(package, configFile))
+                    .AsParallel().ToArray();
 
-            Dictionary<PackageId, ToolPackageException> toolPackageExceptions =
-                new Dictionary<PackageId, ToolPackageException>();
+            Dictionary<RestoredCommandIdentifier, RestoredCommand> succeeded =
```
Comment (dsplaisted) [inline src/dotnet/commands/dotnet-tool/restore/ToolRestoreCommand.cs]: Consider renaming succeeded to downloaded or something.

File: src/dotnet/commands/dotnet-tool/restore/ToolRestoreCommand.cs [hunk 3]
```diff
@@ -96,117 +97,111 @@ public override int Execute()
...
                 _errorReporter.WriteLine(string.Join(
-                                        Environment.NewLine,
-                                        CreateErrorMessage(toolPackageExceptions).Concat(errorMessages)).Red());
+                    Environment.NewLine,
+                    toolRestoreResults.Where(r => !r.IsSuccess).Select(r => r.Message)).Red());
 
                 _reporter.WriteLine(Environment.NewLine);
 
-                _reporter.WriteLine(string.Join(Environment.NewLine, successMessages));
+                _reporter.WriteLine(
```
Comment (dsplaisted) [inline src/dotnet/commands/dotnet-tool/restore/ToolRestoreCommand.cs]: Consider that this will write an extra blank line if there are no success messages.

**Pull request merged**